# Format Core Dataset

A large part of this analysis will be incentives/policies, utility rates, and other more "controllable" factors. I want to get the core dataset together first. 

There are three components to the dataset:
- PV data (aggregated to the ZCTA level)
- Environmental data
- Census data

The environmental data is ready to go. I need to go through the Census data and explore/clean a bit. More than anything I need to prep the PV data so I am going to start with that. I will explore the PV data by itself in the future, right now I am more concerned with prepping it to be part of the larger dataset.

In [1]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## PV Data

PV data has been cleaned and aggregated in `01_prep_pv_data.ipynb`. Read in below and ensure formatted properly.

In [2]:
pv_data = pd.read_csv('../data/pv_systems.csv').drop('Unnamed: 0', axis=1)
pv_data.head()

zcta  num_systems  total_capacity  mean_system_size  median_system_size
0  5001           13          74.705          5.746538               4.200
1  5026            1           5.500          5.500000               5.500
2  5031            4          21.516          5.379000               5.883
3  5032           11          72.133          6.557545               5.400
4  5033           10          71.860          7.186000               7.521

## Census Data

In [3]:
census = pd.read_csv('../data/census.csv').drop('Unnamed: 0', axis=1)
census.head()

zcta    state        lat        long  average_household_income  \
0  85610  Arizona  31.744197  109.722324              53713.747228   
1  85614  Arizona  31.814301  110.919400              67347.031441   
2  85624  Arizona  31.504971  110.692999              56508.955224   
3  85629  Arizona  31.917838  111.019035              91646.185302   
4  85630  Arizona  31.886572  110.181046              57186.339381   

   mean_household_income_lowest_quintile  \
0                                15735.0   
1                                15092.0   
2                                12085.0   
3                                24218.0   
4                                 6123.0   

   mean_household_income_second_quintile  \
0                                28976.0   
1                                33942.0   
2                                26596.0   
3                                55100.0   
4                                16639.0   

   mean_household_income_third_quintile  \
0                               41584.0   
1                               52059.0   
2                               40793.0   
3                               82356.0   
4                               37332.0   

   mean_household_income_fourth_quintile  \
0                                60403.0   
1                                78902.0   
2                                63481.0   
3                               109225.0   
4                                58660.0   

   mean_household_income_highest_quintile  ...  total_population  \
0                                121871.0  ...              1071   
1                                156740.0  ...             23777   
2                                139590.0  ...              1289   
3                                187332.0  ...             25770   
4                                167178.0  ...              1757   

   per_capita_income  median_household_income  household_count  \
0            22831.0                    41422              451   
1            37346.0                    51449            13104   
2            25608.0                    39013              536   
3            30916.0                    82334             8559   
4            30005.0                    38507              937   

   housing_unit_count  housing_unit_occupied_count  housing_unit_median_value  \
0                 691                          451                     118200   
1               16986                        13104                     174800   
2                 860                          536                     240000   
3                9260                         8559                     194200   
4                1226                          937                     163600   

   housing_unit_median_gross_rent  average_years_of_education  \
0                             565                   12.518724   
1                            1011                   14.646702   
2                             719                   14.829431   
3                            1410                   14.201352   
4                             493                   13.923297   

   population_density  
0            2.404896  
1          175.250085  
2            3.291417  
3          111.842558  
4            9.599983  

[5 rows x 88 columns]

In [4]:
census.shape

(15740, 88)

In [5]:
pv_data['zcta'].nunique()

10263

Census data has 15,740 zip codes (ZCTAs) while the PV data only has 10,263. This reflects the number of ZCTAs with no PV systems. It also means the census data will be the core data I use to create the full dataset.

In [6]:
census.isna().sum()

zcta                                0
state                               0
lat                                 0
long                                0
average_household_income          323
                                 ... 
housing_unit_occupied_count         0
housing_unit_median_value           0
housing_unit_median_gross_rent      0
average_years_of_education        204
population_density                  0
Length: 88, dtype: int64

Handling missing data is going to be very interesting with the census data. Imputing will naturally lead to bias (which is dangerous with Census data). But dropping the missing values would result in bias as well (lack of representation). This is something I am going to have to look into quite a bit more. For the purposes of this MVP, I am not going to worry about it too much. 

I am going to leave all of them as missing for now and continue to construct the dataset. Next I want to include the environmental data (`data/zcta_env.json`)

In [7]:
env_df = pd.read_json('../data/zcta_env.json').T.reset_index()
env_df.head()

index    T2M   RH2M  SI_EF_TILTED_SURFACE_HORIZONTAL    PS  CLOUD_AMT_DAY  \
0  86506  11.15  45.83                             5.49  79.0           40.9   
1  86512  11.15  45.83                             5.49  79.0           40.9   
2  85936  11.15  45.83                             5.49  79.0           40.9   
3  86508  11.15  45.83                             5.49  79.0           40.9   
4  87319  11.15  45.83                             5.49  79.0           40.9   

   CDD18_3  HDD18_3  WS2M     TS  FROST_DAYS  TS_AMP  SG_SAA  
0    34.21   228.16  2.93  11.94       10.64   18.21 -100.91  
1    34.21   228.16  2.93  11.94       10.64   18.21 -100.91  
2    34.21   228.16  2.93  11.94       10.64   18.21 -100.91  
3    34.21   228.16  2.93  11.94       10.64   18.21 -100.91  
4    34.21   228.16  2.93  11.94       10.64   18.21 -100.91

Change column names

In [8]:
ENV_COLS = {'index': 'zcta', 'T2M': 'air_temp', 'RH2M': 'relative_humidity', 'SI_EF_TILTED_SURFACE_HORIZONTAL': 'solar_radiation',
            'PS': 'atmospheric_pressure', 'CLOUD_AMT_DAY': 'cloud_amount', 'CDD18_3': 'cooling_degree_days', 'HDD18_3': 'heating_degree_days',
            'WS2M': 'wind_speed', 'TS': 'earth_temp', 'FROST_DAYS': 'frost_days', 'TS_AMP': 'earth_temp_amplitude', 'SG_SAA': 'solar_azimuth_angle'}

In [9]:
env_df = env_df.rename(columns=ENV_COLS)
env_df.head()

zcta  air_temp  relative_humidity  solar_radiation  atmospheric_pressure  \
0  86506     11.15              45.83             5.49                  79.0   
1  86512     11.15              45.83             5.49                  79.0   
2  85936     11.15              45.83             5.49                  79.0   
3  86508     11.15              45.83             5.49                  79.0   
4  87319     11.15              45.83             5.49                  79.0   

   cloud_amount  cooling_degree_days  heating_degree_days  wind_speed  \
0          40.9                34.21               228.16        2.93   
1          40.9                34.21               228.16        2.93   
2          40.9                34.21               228.16        2.93   
3          40.9                34.21               228.16        2.93   
4          40.9                34.21               228.16        2.93   

   earth_temp  frost_days  earth_temp_amplitude  solar_azimuth_angle  
0       11.94       10.64                 18.21              -100.91  
1       11.94       10.64                 18.21              -100.91  
2       11.94       10.64                 18.21              -100.91  
3       11.94       10.64                 18.21              -100.91  
4       11.94       10.64                 18.21              -100.91

In [10]:
env_df.shape

(15755, 13)

Environmental data has even more ZCTAs than the Census data. Still going to use Census data as base dataset. Below I merge the census data and environmental data on the `zcta` column. I perform a `left` join to preserve all observations in the `census` table. Should have 15,740 rows after merge

In [11]:
base_df = census.merge(env_df, on='zcta', how='left')
base_df.head()

zcta    state        lat        long  average_household_income  \
0  85610  Arizona  31.744197  109.722324              53713.747228   
1  85614  Arizona  31.814301  110.919400              67347.031441   
2  85624  Arizona  31.504971  110.692999              56508.955224   
3  85629  Arizona  31.917838  111.019035              91646.185302   
4  85630  Arizona  31.886572  110.181046              57186.339381   

   mean_household_income_lowest_quintile  \
0                                15735.0   
1                                15092.0   
2                                12085.0   
3                                24218.0   
4                                 6123.0   

   mean_household_income_second_quintile  \
0                                28976.0   
1                                33942.0   
2                                26596.0   
3                                55100.0   
4                                16639.0   

   mean_household_income_third_quintile  \
0                               41584.0   
1                               52059.0   
2                               40793.0   
3                               82356.0   
4                               37332.0   

   mean_household_income_fourth_quintile  \
0                                60403.0   
1                                78902.0   
2                                63481.0   
3                               109225.0   
4                                58660.0   

   mean_household_income_highest_quintile  ...  solar_radiation  \
0                                121871.0  ...             5.78   
1                                156740.0  ...             5.75   
2                                139590.0  ...             5.75   
3                                187332.0  ...             5.75   
4                                167178.0  ...             5.78   

   atmospheric_pressure  cloud_amount  cooling_degree_days  \
0                 86.27         32.13                95.20   
1                 90.59         31.87               142.35   
2                 90.59         31.87               142.35   
3                 90.59         31.87               142.35   
4                 86.27         32.13                95.20   

   heating_degree_days  wind_speed  earth_temp  frost_days  \
0               108.36        2.46       17.40         3.3   
1                64.42        2.43       20.51         0.8   
2                64.42        2.43       20.51         0.8   
3                64.42        2.43       20.51         0.8   
4               108.36        2.46       17.40         3.3   

   earth_temp_amplitude  solar_azimuth_angle  
0                 18.02              -100.64  
1                 18.44              -101.18  
2                 18.44              -101.18  
3                 18.44              -101.18  
4                 18.02              -100.64  

[5 rows x 100 columns]

In [12]:
base_df.shape

(15740, 100)

The last thing I need to do is add PV data and to `base_df`

In [13]:
pv_data.head()

zcta  num_systems  total_capacity  mean_system_size  median_system_size
0  5001           13          74.705          5.746538               4.200
1  5026            1           5.500          5.500000               5.500
2  5031            4          21.516          5.379000               5.883
3  5032           11          72.133          6.557545               5.400
4  5033           10          71.860          7.186000               7.521

In [14]:
base_df['zcta'].dtype

dtype('int64')

In [15]:
pv_data['zcta'].dtype

dtype('int64')

In [16]:
base_df = base_df.merge(pv_data, on='zcta', how='left')
base_df.head()

zcta    state        lat        long  average_household_income  \
0  85610  Arizona  31.744197  109.722324              53713.747228   
1  85614  Arizona  31.814301  110.919400              67347.031441   
2  85624  Arizona  31.504971  110.692999              56508.955224   
3  85629  Arizona  31.917838  111.019035              91646.185302   
4  85630  Arizona  31.886572  110.181046              57186.339381   

   mean_household_income_lowest_quintile  \
0                                15735.0   
1                                15092.0   
2                                12085.0   
3                                24218.0   
4                                 6123.0   

   mean_household_income_second_quintile  \
0                                28976.0   
1                                33942.0   
2                                26596.0   
3                                55100.0   
4                                16639.0   

   mean_household_income_third_quintile  \
0                               41584.0   
1                               52059.0   
2                               40793.0   
3                               82356.0   
4                               37332.0   

   mean_household_income_fourth_quintile  \
0                                60403.0   
1                                78902.0   
2                                63481.0   
3                               109225.0   
4                                58660.0   

   mean_household_income_highest_quintile  ...  heating_degree_days  \
0                                121871.0  ...               108.36   
1                                156740.0  ...                64.42   
2                                139590.0  ...                64.42   
3                                187332.0  ...                64.42   
4                                167178.0  ...               108.36   

   wind_speed  earth_temp  frost_days  earth_temp_amplitude  \
0        2.46       17.40         3.3                 18.02   
1        2.43       20.51         0.8                 18.44   
2        2.43       20.51         0.8                 18.44   
3        2.43       20.51         0.8                 18.44   
4        2.46       17.40         3.3                 18.02   

   solar_azimuth_angle  num_systems  total_capacity  mean_system_size  \
0              -100.64         13.0          70.150          5.396154   
1              -101.18       1012.0        7015.507          6.932319   
2              -101.18         24.0         150.860          6.285833   
3              -101.18       1186.0        8934.678          7.533455   
4              -100.64         37.0         258.010          6.973243   

   median_system_size  
0              5.3800  
1              5.9850  
2              5.8650  
3              7.2975  
4              6.4800  

[5 rows x 104 columns]

In [17]:
base_df.shape

(16063, 104)

A few rows were added after the merge

In [18]:
# check duplicates
base_df.duplicated(subset=['zcta']).sum()

323

323 duplicated ZCTA values

In [19]:
duplicate_zctas = base_df[base_df.duplicated(subset=['zcta'])]['zcta']

In [20]:
base_df[base_df['zcta'].isin(duplicate_zctas)]

zcta        state        lat       long  average_household_income  \
2169   6470  Connecticut  41.395083  73.317663             152776.476561   
2170   6470  Connecticut  41.395083  73.317663             152776.476561   
2171   6472  Connecticut  41.382766  72.775194             129469.583333   
2172   6472  Connecticut  41.382766  72.775194             129469.583333   
2173   6480  Connecticut  41.598834  72.589071             121043.197006   
...     ...          ...        ...        ...                       ...   
14678  5452      Vermont  44.538872  73.050085             102689.933491   
14682  5477      Vermont  44.416769  72.962422             107859.183673   
14683  5477      Vermont  44.416769  72.962422             107859.183673   
14685  5495      Vermont  44.429068  73.096245             111245.292560   
14686  5495      Vermont  44.429068  73.096245             111245.292560   

       mean_household_income_lowest_quintile  \
2169                                 30900.0   
2170                                 30900.0   
2171                                 41192.0   
2172                                 41192.0   
2173                                 21074.0   
...                                      ...   
14678                                21305.0   
14682                                29507.0   
14683                                29507.0   
14685                                21745.0   
14686                                21745.0   

       mean_household_income_second_quintile  \
2169                                 76312.0   
2170                                 76312.0   
2171                                 76412.0   
2172                                 76412.0   
2173                                 57895.0   
...                                      ...   
14678                                53820.0   
14682                                65140.0   
14683                                65140.0   
14685                                55090.0   
14686                                55090.0   

       mean_household_income_third_quintile  \
2169                               125581.0   
2170                               125581.0   
2171                               100063.0   
2172                               100063.0   
2173                                97127.0   
...                                     ...   
14678                               85352.0   
14682                               99379.0   
14683                               99379.0   
14685                               91120.0   
14686                               91120.0   

       mean_household_income_fourth_quintile  \
2169                                186743.0   
2170                                186743.0   
2171                                139226.0   
2172                                139226.0   
2173                                139356.0   
...                                      ...   
14678                               119009.0   
14682                               132808.0   
14683                               132808.0   
14685                               134183.0   
14686                               134183.0   

       mean_household_income_highest_quintile  ...  heating_degree_days  \
2169                                 344346.0  ...               232.55   
2170                                 344346.0  ...               232.55   
2171                                 290454.0  ...               232.55   
2172                                 290454.0  ...               232.55   
2173                                 289765.0  ...               305.34   
...                                       ...  ...                  ...   
14678                                233965.0  ...               358.04   
14682                                212461.0  ...               356.94   
14683                                212461.0  ...               356.94   
14685                                254088.0  .

Drop duplicated ZCTAs. Should have 15,740 rows

In [21]:
base_df = base_df.drop_duplicates(subset=['zcta']).reset_index(drop=True)
base_df.head()

zcta    state        lat        long  average_household_income  \
0  85610  Arizona  31.744197  109.722324              53713.747228   
1  85614  Arizona  31.814301  110.919400              67347.031441   
2  85624  Arizona  31.504971  110.692999              56508.955224   
3  85629  Arizona  31.917838  111.019035              91646.185302   
4  85630  Arizona  31.886572  110.181046              57186.339381   

   mean_household_income_lowest_quintile  \
0                                15735.0   
1                                15092.0   
2                                12085.0   
3                                24218.0   
4                                 6123.0   

   mean_household_income_second_quintile  \
0                                28976.0   
1                                33942.0   
2                                26596.0   
3                                55100.0   
4                                16639.0   

   mean_household_income_third_quintile  \
0                               41584.0   
1                               52059.0   
2                               40793.0   
3                               82356.0   
4                               37332.0   

   mean_household_income_fourth_quintile  \
0                                60403.0   
1                                78902.0   
2                                63481.0   
3                               109225.0   
4                                58660.0   

   mean_household_income_highest_quintile  ...  heating_degree_days  \
0                                121871.0  ...               108.36   
1                                156740.0  ...                64.42   
2                                139590.0  ...                64.42   
3                                187332.0  ...                64.42   
4                                167178.0  ...               108.36   

   wind_speed  earth_temp  frost_days  earth_temp_amplitude  \
0        2.46       17.40         3.3                 18.02   
1        2.43       20.51         0.8                 18.44   
2        2.43       20.51         0.8                 18.44   
3        2.43       20.51         0.8                 18.44   
4        2.46       17.40         3.3                 18.02   

   solar_azimuth_angle  num_systems  total_capacity  mean_system_size  \
0              -100.64         13.0          70.150          5.396154   
1              -101.18       1012.0        7015.507          6.932319   
2              -101.18         24.0         150.860          6.285833   
3              -101.18       1186.0        8934.678          7.533455   
4              -100.64         37.0         258.010          6.973243   

   median_system_size  
0              5.3800  
1              5.9850  
2              5.8650  
3              7.2975  
4              6.4800  

[5 rows x 104 columns]

In [22]:
base_df.shape

(15740, 104)

Save core dataset as `data/base_data.csv`

In [23]:
base_df.to_csv('../data/base_data.csv')